src/Compilers/Core/MSBuildTask/MapSourceRoots.cs

```diff
@@ -0,0 +1,232 @@
+        }
+
+        private static string EndWithSlash(string path)
+            => (path[path.Length - 1] == '/') ? path : path + '/';
+
+        private static bool EndsWithDirectorySeparator(string path)
+        {
+            if (path.Length == 0)
+            {
+                return false;
+            }
+
+            char c = path[path.Length - 1];
+            return c == Path.DirectorySeparatorChar || c == Path.AltDirectorySeparatorChar;
+        }
+
+        private void MergeSourceRootMetadata(ITaskItem left, ITaskItem right)
+        {
+            foreach (string metadataName in right.MetadataNames)
+            {
+                var leftValue = left.GetMetadata(metadataName);
+                var rightValue = right.GetMetadata(metadataName);
+
+                if (!string.IsNullOrEmpty(leftValue) && !string.IsNullOrEmpty(rightValue))
+                {
+                    Log.LogErrorFromResources("MapSourceRoots.ContainsDuplicate", Names.SourceRoot, right.ItemSpec, metadataName, leftValue, rightValue);
+                }
+                else if (string.IsNullOrEmpty(leftValue) && !string.IsNullOrEmpty(rightValue))
+                {
+                    left.SetMetadata(metadataName, rightValue);
+                }
+            }
+        }
+
+        public override bool Execute()
+        {
+            // Merge metadata of SourceRoot items with the same identity.
+            var mappedSourceRoots = new List<ITaskItem>();
+            var rootByItemSpec = new Dictionary<string, ITaskItem>();
+            foreach (var sourceRoot in SourceRoots)
+            {
+                // The SourceRoot is required to have a trailing directory separator.
+                // We do not append one implicitly as we do not know which separator to append on Windows.
+                // The usage of SourceRoot might be sensitive to what kind of separator is used (e.g. in SourceLink where it needs
+                // to match the corresponding separators used in paths given to the compiler).
+                if (!EndsWithDirectorySeparator(sourceRoot.ItemSpec))
+                {
+                    Log.LogErrorFromResources("MapSourceRoots.PathMustEndWithSlashOrBackslash", Names.SourceRoot, sourceRoot.ItemSpec);
+                }
+
+                if (rootByItemSpec.TryGetValue(sourceRoot.ItemSpec, out var existingRoot))
+                {
+                    ReportConflictingWellKnownMetadata(existingRoot, sourceRoot);
+                    sourceRoot.CopyMetadataTo(existingRoot);
+                }
+                else
+                {
+                    rootByItemSpec.Add(sourceRoot.ItemSpec, sourceRoot);
+                    mappedSourceRoots.Add(sourceRoot);
+                }
+            }
+
+            if (Log.HasLoggedErrors)
+            {
+                return false;
+            }
+
+            if (Deterministic)
+            {
+                var topLevelMappedPaths = new Dictionary<string, string>();
```

Comment:
>var topLevelMappedPaths = new Dictionary<string, string>(); [](start = 16, length = 59)

Why are we using ordinal path comparison here? The compiler is pretty inconsistent here but generally favors insensiitive.

Reply:
These paths are generally produced by a source control package (e.g. one based on GitLib library). It is expected that the tool produces paths that have consistent casing. This particular dictionary is used to map submodule paths to their containing source roots. When the submodule paths are produced by the GitLib library they should match the casing exactly.